src/rust/engine/fs/store/src/remote/reapi_tests.rs

```diff
@@ -167,14 +163,140 @@ fn assert_cas_store(
   }
 }
 
+fn store_source(data: Bytes) -> StoreSource {
+  Box::new(std::io::Cursor::new(data))
+}
+
+#[tokio::test]
+async fn store_one_chunk() {
+  let testdata = TestData::roland();
+  let cas = StubCAS::empty();
+  let provider = new_provider(&cas).await;
+
+  provider
+    .store(testdata.digest(), store_source(testdata.bytes()))
+    .await
+    .unwrap();
+
+  assert_cas_store(&cas, testdata.fingerprint(), testdata.bytes(), 1, 1024)
+}
+#[tokio::test]
+async fn store_multiple_chunks() {
+  let cas = StubCAS::empty();
+  let chunk_size = 10 * 1024;
+  let provider = Provider::new(remote_options(
+    cas.address(),
+    chunk_size,
+    0, // disable batch API, force streaming API
+  ))
+  .await
+  .unwrap();
+
+  let all_the_henries = big_file_bytes();
+  let fingerprint = big_file_fingerprint();
+  let digest = Digest::new(fingerprint, all_the_henries.len());
+
+  provider
+    .store(digest, store_source(all_the_henries.clone()))
+    .await
+    .unwrap();
+
+  assert_cas_store(&cas, fingerprint, all_the_henries, 98, chunk_size)
+}
+
+#[tokio::test]
+async fn store_empty_file() {
+  let testdata = TestData::empty();
+  let cas = StubCAS::empty();
+  let provider = new_provider(&cas).await;
+
+  provider
+    .store(testdata.digest(), store_source(testdata.bytes()))
+    .await
+    .unwrap();
+
+  assert_cas_store(&cas, testdata.fingerprint(), testdata.bytes(), 1, 1024)
+}
+
+#[tokio::test]
+async fn store_grpc_error() {
+  let testdata = TestData::roland();
+  let cas = StubCAS::cas_always_errors();
+  let provider = new_provider(&cas).await;
+
+  let error = provider
+    .store(testdata.digest(), store_source(testdata.bytes()))
+    .await
+    .expect_err("Want err");
+  assert!(
+    error.contains("StubCAS is configured to always fail"),
+    "Bad error message, got: {error}"
+  );
+}
+
+#[tokio::test]
+async fn store_connection_error() {
+  let testdata = TestData::roland();
+  let provider = Provider::new(remote_options(
+    "http://doesnotexist.example".to_owned(),
+    10 * MEGABYTES,
+    crate::tests::STORE_BATCH_API_SIZE_LIMIT,
+  ))
+  .await
+  .unwrap();
+
+  let error = provider
+    .store(testdata.digest(), store_source(testdata.bytes()))
+    .await
+    .expect_err("Want err");
+  assert!(
+    error.contains("Unavailable: \"error trying to connect: dns error"),
+    "Bad error message, got: {error}"
+  );
+}
+
+#[tokio::test]
+async fn store_source_read_error_immediately() {
```

Comment:
Most of these tests are just copied/adapted from the `store_bytes` ones below, except for this one and `store_source_read_error_later` which test the new/interesting code path: what happens if the `AsyncRead`s fail?